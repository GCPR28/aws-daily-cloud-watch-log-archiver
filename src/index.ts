import * as crypto from 'crypto';
import { SecureBucket, SecureBucketEncryption } from '@yicr/secure-bucket';
import * as cdk from 'aws-cdk-lib';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as scheduler from 'aws-cdk-lib/aws-scheduler';
import { Construct } from 'constructs';
import { LogArchiverFunction } from './log-archiver-function';

export interface DailyCloudWatchLogArchiverProps {
  readonly schedule: ScheduleProperty;
}

export interface ScheduleProperty {
  readonly name: string;
  readonly description: string;
  readonly target: ScheduleTargetProperty;
}

export interface ScheduleTargetProperty {
  readonly logGroupName: string;
  readonly destinationPrefix: string;
}

export class DailyCloudWatchLogArchiver extends Construct {
  constructor(scope: Construct, id: string, props: DailyCloudWatchLogArchiverProps) {
    super(scope, id);

    // ðŸ‘‡Get current account & region
    // const account = cdk.Stack.of(this).account;
    const region = cdk.Stack.of(this).region;

    const randomNameKey = crypto.createHash('shake256', { outputLength: 4 })
      .update(cdk.Names.uniqueId(this))
      .digest('hex');

    // ðŸ‘‡Create Backup S3 Bucket
    const logArchiveBucket = new SecureBucket(this, 'LogArchiveBucket', {
      bucketName: `log-archive-${randomNameKey}`,
      encryption: SecureBucketEncryption.KMS_MANAGED,
    });
    logArchiveBucket.addToResourcePolicy(new iam.PolicyStatement({
      effect: iam.Effect.ALLOW,
      principals: [
        new iam.ServicePrincipal(`logs.${region}.amazonaws.com`),
      ],
      actions: [
        's3:GetBucketAcl',
      ],
      resources: [
        logArchiveBucket.bucketArn,
      ],
    }));
    logArchiveBucket.addToResourcePolicy(new iam.PolicyStatement({
      effect: iam.Effect.ALLOW,
      principals: [
        new iam.ServicePrincipal(`logs.${region}.amazonaws.com`),
      ],
      actions: [
        's3:PutObject',
      ],
      resources: [
        `${logArchiveBucket.bucketArn}/*`,
      ],
      conditions: {
        StringEquals: {
          's3:x-amz-acl': 'bucket-owner-full-control',
        },
      },
    }));

    // ðŸ‘‡Create Lambda Execution role.
    const lambdaExecutionRole = new iam.Role(this, 'LambdaExecutionRole', {
      roleName: `daily-cw-log-archiver-lambda-exec-${randomNameKey}-role`,
      description: '',
      assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
      managedPolicies: [
        iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSLambdaBasicExecutionRole'),
      ],
      inlinePolicies: {
        ['log-export-policy']: new iam.PolicyDocument({
          statements: [
            new iam.PolicyStatement({
              effect: iam.Effect.ALLOW,
              actions: [
                'logs:CreateExportTask',
              ],
              resources: ['*'],
            }),
          ],
        }),
        ['put-bucket-policy']: new iam.PolicyDocument({
          statements: [
            new iam.PolicyStatement({
              effect: iam.Effect.ALLOW,
              actions: [
                's3:GetBucketAcl',
                's3:PutObject',
              ],
              resources: [
                logArchiveBucket.bucketArn,
              ],
            }),
          ],
        }),
      },
    });

    // ðŸ‘‡Create Lambda Function
    const lambdaFunction = new LogArchiverFunction(this, 'LogArchiveFunction', {
      functionName: `daily-cw-log-archiver-${randomNameKey}-func`,
      description: 'A function to archive logs s3 bucket from CloudWatch Logs.',
      environment: {
        BUCKET_NAME: logArchiveBucket.bucketName,
      },
      role: lambdaExecutionRole,
    });

    // ðŸ‘‡EventBridge Scheduler IAM Role
    const schedulerExecutionRole = new iam.Role(this, 'SchedulerExecutionRole', {
      assumedBy: new iam.ServicePrincipal('scheduler.amazonaws.com'),
      inlinePolicies: {
        ['lambda-invoke-policy']: new iam.PolicyDocument({
          statements: [
            new iam.PolicyStatement({
              effect: iam.Effect.ALLOW,
              actions: [
                'lambda:InvokeFunction',
              ],
              resources: [
                lambdaFunction.functionArn,
                `${lambdaFunction.functionArn}:*`,
              ],
            }),
          ],
        }),
      },
    });

    // ðŸ‘‡Create Schedule Group
    const scheduleGroup = new scheduler.CfnScheduleGroup(this, 'ScheduleGroup', {
      name: `log-archive-schedule-${randomNameKey}-group`,
    });

    // ðŸ‘‡Schedule ID prefix
    const idPrefix = crypto.createHash('shake256', { outputLength: 4 })
      .update(props.schedule.name)
      .digest('hex');

    // ðŸ‘‡Schedule
    new scheduler.CfnSchedule(this, `Schedule${idPrefix}`, {
      name: props.schedule.name,
      description: props.schedule.description,
      state: 'ENABLED',
      groupName: scheduleGroup.name,
      flexibleTimeWindow: {
        mode: 'OFF',
      },
      scheduleExpressionTimezone: 'UTC',
      scheduleExpression: 'cron(* 13 * * ? *)', // todo: when loop to increment minute
      target: {
        arn: lambdaFunction.functionArn,
        roleArn: schedulerExecutionRole.roleArn,
        input: JSON.stringify({
          logGroupName: props.schedule.target.logGroupName,
          destinationPrefix: props.schedule.target.destinationPrefix,
        }),
        retryPolicy: {
          maximumEventAgeInSeconds: 60,
          maximumRetryAttempts: 0,
        },
      },
    });
  }
}