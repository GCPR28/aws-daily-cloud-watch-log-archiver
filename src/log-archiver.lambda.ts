import { CloudWatchLogsClient, CreateExportTaskCommand } from '@aws-sdk/client-cloudwatch-logs';
import { Context } from 'aws-lambda';

export class EnvironmentVariableError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'EnvironmentVariableError';
  }
}

export class InputVariableError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'InputVariableError';
  }
}

export interface EventInput {
  readonly logGroupName?: string;
  readonly destinationPrefix?: string;
}

const cwLogsClient = new CloudWatchLogsClient({});

export const handler = async (event: EventInput, context: Context): Promise<string | undefined> => {
  console.log(`Event: ${JSON.stringify(event, null, 2)}`);
  console.log(`Context: ${JSON.stringify(context, null, 2)}`);

  // do validation
  if (!process.env.BucketName) {
    throw new EnvironmentVariableError('BucketName environment variable not set.');
  }
  if (!event.logGroupName) {
    throw new InputVariableError('event input logGroupName environment variable not set.');
  }
  if (!event.destinationPrefix) {
    throw new InputVariableError('event input destinationPrefix variable not set.');
  }
  const now = new Date();
  const targetFromTime = new Date(now.getFullYear(), now.getMonth(), now.getDate(), 0, 0, 0, 0).getTime() - (1000 * 60 * 60 * 24);
  const targetToTime = targetFromTime + (1000 * 60 * 60 * 24);
  const targetDate = new Date(targetFromTime);
  var y = targetDate.getFullYear();
  var m = ('00' + (targetDate.getMonth()+1)).slice(-2);
  var d = ('00' + (targetDate.getDate())).slice(-2);

  const params = {
    destination: process.env.BucketName,
    logGroupName: process.env.LogGroupName,
    from: targetFromTime,
    to: targetToTime,
    destinationPrefix: `${event.destinationPrefix}/${y}-${m}-${d}`,
  };
  console.log(`CommandParams: ${JSON.stringify(params, null, 2)}`);

  const result = await cwLogsClient.send(new CreateExportTaskCommand(params));
  console.log(`CommandResult: ${JSON.stringify(result, null, 2)}`);

  return result.taskId;
};

