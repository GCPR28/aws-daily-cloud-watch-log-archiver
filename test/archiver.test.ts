import { App, Stack } from 'aws-cdk-lib';
import { Match, Template } from 'aws-cdk-lib/assertions';
import { DailyCloudWatchLogArchiver } from '../src';

describe('DailyCloudWatchLogArchiver Testing', () => {
  const app = new App();
  const stack = new Stack(app, 'TestingStack', {
    env: {
      account: '123456789012',
      region: 'us-east-1',
    },
  });

  new DailyCloudWatchLogArchiver(stack, 'DailyCloudWatchLogArchiver', {
    schedules: [
      {
        name: 'example-log-archive-1st-rule',
        description: 'example log archive 1st rule.',
        target: {
          logGroupName: 'example-log-1st-group',
          destinationPrefix: 'example-1st-log',
        },
      },
      {
        name: 'example-log-archive-2nd-rule',
        description: 'example log archive 2nd rule.',
        target: {
          logGroupName: 'example-log-2nd-group',
          destinationPrefix: 'example-2nd-log',
        },
      },
    ],
  });

  const template = Template.fromStack(stack);

  describe('Bucket Testing', () => {

    it('Should have bucket encryption', () => {
      template.hasResourceProperties('AWS::S3::Bucket', {
        BucketEncryption: Match.objectEquals({
          ServerSideEncryptionConfiguration: [
            {
              ServerSideEncryptionByDefault: {
                SSEAlgorithm: 'aws:kms',
              },
            },
          ],
        }),
      });
    });

    it('Should have bucket resource policy', () => {
      template.hasResourceProperties('AWS::S3::BucketPolicy', {
        PolicyDocument: {
          Version: '2012-10-17',
          Statement: Match.arrayWith([
            Match.objectEquals({
              Effect: 'Allow',
              Action: 's3:GetBucketAcl',
              Principal: {
                Service: 'logs.us-east-1.amazonaws.com',
              },
              Resource: {
                'Fn::GetAtt': [
                  Match.stringLikeRegexp('DailyCloudWatchLogArchiverLogArchiveBucket'),
                  'Arn',
                ],
              },
            }),
            Match.objectEquals({
              Effect: 'Allow',
              Action: 's3:PutObject',
              Principal: {
                Service: 'logs.us-east-1.amazonaws.com',
              },
              Resource: {
                'Fn::Join': [
                  '',
                  [
                    {
                      'Fn::GetAtt': [
                        Match.stringLikeRegexp('DailyCloudWatchLogArchiverLogArchiveBucket'),
                        'Arn',
                      ],
                    },
                    '/*',
                  ],
                ],
              },
              Condition: {
                StringEquals: {
                  's3:x-amz-acl': 'bucket-owner-full-control',
                },
              },
            }),
          ]),
        },
      });
    });

  });

  describe('Lambda Testing', () => {

    it('Should have lambda execution role', () => {
      template.hasResourceProperties('AWS::IAM::Role', Match.objectEquals({
        RoleName: Match.stringLikeRegexp('daily-cw-log-archiver-lambda-exec-.*-role'),
        AssumeRolePolicyDocument: Match.objectEquals({
          Version: '2012-10-17',
          Statement: Match.arrayWith([
            Match.objectEquals({
              Effect: 'Allow',
              Principal: {
                Service: 'lambda.amazonaws.com',
              },
              Action: 'sts:AssumeRole',
            }),
          ]),
        }),
        ManagedPolicyArns: Match.arrayWith([
          {
            'Fn::Join': Match.arrayEquals([
              '',
              Match.arrayEquals([
                'arn:',
                {
                  Ref: 'AWS::Partition',
                },
                ':iam::aws:policy/service-role/AWSLambdaBasicExecutionRole',
              ]),
            ]),
          },
        ]),
        Policies: Match.arrayEquals([
          {
            PolicyName: 'log-export-policy',
            PolicyDocument: Match.objectEquals({
              Version: '2012-10-17',
              Statement: [
                Match.objectEquals({
                  Effect: 'Allow',
                  Action: 'logs:CreateExportTask',
                  Resource: '*',
                }),
              ],
            }),
          },
          {
            PolicyName: 'put-bucket-policy',
            PolicyDocument: Match.objectEquals({
              Version: '2012-10-17',
              Statement: [
                Match.objectEquals({
                  Effect: 'Allow',
                  Action: Match.arrayEquals([
                    's3:GetBucketAcl',
                    's3:PutObject',
                  ]),
                  Resource: {
                    'Fn::GetAtt': Match.arrayEquals([
                      Match.stringLikeRegexp('DailyCloudWatchLogArchiverLogArchiveBucket.*'),
                      'Arn',
                    ]),
                  },
                }),
              ],
            }),
          },
        ]),
      }));
    });

    it('Should have lambda function', () => {
      template.hasResourceProperties('AWS::Lambda::Function', Match.objectEquals({
        FunctionName: Match.stringLikeRegexp('daily-cw-log-archiver-.*-func'),
        Handler: 'index.handler',
        Runtime: 'nodejs18.x',
        Code: {
          S3Bucket: Match.anyValue(),
          S3Key: Match.stringLikeRegexp('.*.zip'),
        },
        Description: Match.anyValue(),
        Environment: {
          Variables: {
            AWS_NODEJS_CONNECTION_REUSE_ENABLED: '1',
            BUCKET_NAME: {
              Ref: Match.stringLikeRegexp('DailyCloudWatchLogArchiverLogArchiveBucket.*'),
            },
          },
        },
        Role: {
          'Fn::GetAtt': [
            Match.stringLikeRegexp('DailyCloudWatchLogArchiverLambdaExecutionRole.*'),
            'Arn',
          ],
        },
      }));
    });
  });

  // todo: scheduler property.
  it('Should have Schedule', () => {
    template.hasResourceProperties('AWS::Scheduler::ScheduleGroup', Match.objectEquals({
      Name: Match.stringLikeRegexp('log-archive-schedule-.*-group'),
    }));
    template.hasResourceProperties('AWS::Scheduler::Schedule', Match.objectEquals({
      Name: Match.anyValue(),
      Description: Match.anyValue(),
      GroupName: Match.stringLikeRegexp('log-archive-schedule-.*-group'),
      State: 'ENABLED',
      FlexibleTimeWindow: {
        Mode: 'OFF',
      },
      ScheduleExpressionTimezone: 'UTC',
      ScheduleExpression: Match.stringLikeRegexp('cron(.* 13 * * ? *)'),
      Target: Match.objectEquals({
        Arn: {
          'Fn::GetAtt': [
            Match.stringLikeRegexp('DailyCloudWatchLogArchiverLogArchiveFunction.*'),
            'Arn',
          ],
        },
        RoleArn: {
          'Fn::GetAtt': [
            Match.stringLikeRegexp('DailyCloudWatchLogArchiverSchedulerExecutionRole.*'),
            'Arn',
          ],
        },
        Input: Match.stringLikeRegexp('{"logGroupName":"example-log-.*-group","destinationPrefix":"example-.*-log"}'),
        RetryPolicy: {
          MaximumEventAgeInSeconds: 60,
          MaximumRetryAttempts: 0,
        },
      }),
    }));
    template.resourceCountIs('AWS::Scheduler::Schedule', 2);
  });

  it('Should match snapshot', () => {
    expect(template.toJSON()).toMatchSnapshot('archiver');
  });
});
