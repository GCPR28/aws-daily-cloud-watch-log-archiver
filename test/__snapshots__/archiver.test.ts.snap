// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DailyCloudWatchLogArchiver Testing Should match snapshot: archiver 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "DailyCloudWatchLogArchiverLambdaExecutionRoleFCCBB526": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": "logs:CreateExportTask",
                  "Effect": "Allow",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "log-export-policy",
          },
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "s3:GetBucketAcl",
                    "s3:PutObject",
                  ],
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::GetAtt": Array [
                      "DailyCloudWatchLogArchiverLogArchiveBucketDED235D5",
                      "Arn",
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "put-bucket-policy",
          },
        ],
        "RoleName": "daily-cw-log-archiver-lambda-exec-c08c1234-role",
      },
      "Type": "AWS::IAM::Role",
    },
    "DailyCloudWatchLogArchiverLogArchiveBucketDED235D5": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AccessControl": "Private",
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "aws:kms",
              },
            },
          ],
        },
        "BucketName": "log-archive-c08c1234",
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "VersioningConfiguration": Object {
          "Status": "Enabled",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "DailyCloudWatchLogArchiverLogArchiveBucketPolicy6184475E": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "DailyCloudWatchLogArchiverLogArchiveBucketDED235D5",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:*",
              "Condition": Object {
                "Bool": Object {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": Object {
                "AWS": "*",
              },
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "DailyCloudWatchLogArchiverLogArchiveBucketDED235D5",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "DailyCloudWatchLogArchiverLogArchiveBucketDED235D5",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "s3:GetBucketAcl",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "logs.us-east-1.amazonaws.com",
              },
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "DailyCloudWatchLogArchiverLogArchiveBucketDED235D5",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": "s3:PutObject",
              "Condition": Object {
                "0": Object {
                  "StringEquals": Object {
                    "s3:x-amz-acl": "bucket-owner-full-control",
                  },
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "Service": "logs.us-east-1.amazonaws.com",
              },
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "DailyCloudWatchLogArchiverLogArchiveBucketDED235D5",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "DailyCloudWatchLogArchiverLogArchiveFunctionF76E92C4": Object {
      "DependsOn": Array [
        "DailyCloudWatchLogArchiverLambdaExecutionRoleFCCBB526",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": "cdk-hnb659fds-assets-123456789012-us-east-1",
          "S3Key": "HASH.zip",
        },
        "Description": "A function to archive logs s3 bucket from CloudWatch Logs.",
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "FunctionName": "daily-cw-log-archiver-c08c1234-func",
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "DailyCloudWatchLogArchiverLambdaExecutionRoleFCCBB526",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
